### ISSUE TITLE
[C API] while loop: unable to access operations defined outside of the loop from within the loop

### ISSUE THREAD

**Issue by melissagrueter** [tensorflow/tensorflow#28195]
*System information*

* Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
* OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Mojave 10.14.4
* Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
* TensorFlow installed from (source or binary): source
* TensorFlow version (use command below): master
* Python version: Python 3.7.3
* Bazel version (if compiling from source): 0.24.1
* GCC/Compiler version (if compiling from source): 
* CUDA/cuDNN version: N/A
* GPU model and memory: N/A

*Describe the current behavior*

I am unable to access operations defined outside of the while loop from within the loop. 

The C API while loop creates separate conditional and body graphs, so an error is thrown when we try to use operations defined in the outer graph within the body graph. See earlier discussion with @skye [here](https://github.com/tensorflow/tensorflow/issues/26371#issuecomment-482358762).

*Describe the expected behavior*

I would expect the behavior of the C API while loop to match the Python API while loop, where accessing operations defined outside the while loop works. 

I've included a minimal working example in Python that demonstrates the expected behavior below. In this example, we are able to use “increment” in the loop body even though it's defined outside the loop.

```
import tensorflow as tf

increment = tf.constant(1, name='one')

def loop_cond(loop_var):
    return tf.math.less(loop_var, 10)

def loop_body(loop_var):
    return loop_var + increment

loop_input = tf.Variable(0, name='loop_input')

loop_output = tf.while_loop(loop_cond, loop_body, [loop_input])

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    print(sess.run(loop_input)) // should be 0
    print(sess.run(loop_output)) // should be 10 
```

*Code to reproduce the issue*

I replicated the example above as a unit test in while_loop_test.cc. Once again, we try to use “increment” within the loop body, but here we get an error since it's not part of the body graph. The error message is “Node 'add': Unknown input node 'scalar'”. 

```
TEST_F(CApiWhileLoopTest, AccessOuterOp) {
  Init(1);
  // increment = 1
  // while (i < 10) {
  //   i = i + increment
  // }

  // Create increment *in the outer graph*
  TF_Operation* increment = ScalarConst(1, graph_, s_);

  // Create cond graph: i < 10 
  TF_Operation* ten = ScalarConst(10, params_->cond_graph, s_);
  TF_Operation* less_than = LessThan(params_->cond_inputs[0], {ten, 0}, params_->cond_graph, s_);
  DCHECK_EQ(TF_OK, TF_GetCode(s_)) << TF_Message(s_);
  params_->cond_output = {less_than, 0};

  // Create body graph: i = i + increment
  TF_Operation* add = Add(params_->body_inputs[0], {increment, 0}, params_->body_graph, s_);
  ASSERT_EQ(TF_OK, TF_GetCode(s_)) << TF_Message(s_);
  params_->body_outputs[0] = {add, 0};

  ExpectOK();
}
```
This test can be copy-pasted into while_loop_test.cc and run with the following command:
 bazel run //tensorflow/c:while_loop_test. 

*Other info / logs*
The above examples are intended to be as minimal as possible, so they're not practical. However, accessing outside operations would be important when updating external variables, for example when training within a loop (we would need to update external weights and biases).

We discovered this issue while using TF Java, after exposing the C API while loop to Java in [this commit](https://github.com/tensorflow/tensorflow/commit/6799f6e990f856452c467f31448c211fce94bbc1).

**Comment by melissagrueter** [comment]
cc @saxenasaurabh

**Comment by melissagrueter** [comment]
In my earlier discussion with Skye about this issue (linked above), we discussed capturing as a solution to this problem:

“Based on some offline discussion, I think we're leaning towards adding the ability to "capture" Tensors in a TF_Graph. This is very similar to allowing inputs from other Graphs. This would be a step towards representing functions as graphs in the C API, which is probably a prerequisite for while_v2, so it'd be nice to move towards that final goal.” - Skye

@saxenasaurabh, is that still what the TF team has in mind? If so, would that necessitate implementing a version of Python's FuncGraphs for the C API, or is there an alternative approach you'd suggest? Of course, I'd like to consider the simplest solutions first. 

Thanks!

**Comment by skye** [comment]
Hi Melissa, sorry for the delay. I should have looked into this earlier, but I just noticed that the "Unknown input node" message comes from importing a GraphDef: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/graph/graph_constructor.cc#L561

I'm guessing this means that the body graph can somehow already reference nodes from the outer graph (maybe because we usually check this in Python?), but then it complains when we try to import the body graph into the outer graph:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.cc#L2430
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.cc#L2282

A quick fix might be to just add _all_ the nodes in the outer graph to the input_map of that import, not just the loop inputs (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.cc#L2266). Melissa, do you wanna give this a try and see if it works, or if we get more errors? There might be performance issues down the line creating such a large input_map, but we can start with this just to see if this is the only problem. Please let me know if you need more help or context.

**Comment by melissagrueter** [comment]
Hi @skye,

Thank you for your super helpful (as always) insights! 

Just a couple things I wanted to clarify:

I thought that the input_map maps nodes in the src graph that we'd like to replace to their replacements in the dst graph. In the existing code, src_inputs are input_map's keys and dst_inputs are the corresponding values, because we want to replace the loop input nodes created in the cond and body graph with the loop input nodes created in the outer graph. See comment [here](https://github.com/tensorflow/tensorflow/blob/e9bd33a41d181593748d12af950f35381e9e530e/tensorflow/c/c_api.cc#L2244): “Any node in src_graph with input src_inputs[i] will have that input replaced with dst_inputs[i].”  Since in this case we don't actually want to replace any nodes, were you suggesting adding every node in the outer graph to input_map as both a key and its own value? (In order to avoid triggering [this condition](https://github.com/tensorflow/tensorflow/blob/9d558ddba177859ebd7e4018a405f68dcc5e1836/tensorflow/core/graph/graph_constructor.cc#L555) for any input from the outer graph.) Let me know if I've misunderstood any of the above.

Also, for additional context if at all useful/interesting, this is actually sort of similar to the hack I tried earlier when trying to enable similar functionality using control dependencies  — except I added the control dependencies I wanted to opts.control_dependencies [here](https://github.com/tensorflow/tensorflow/blob/e9bd33a41d181593748d12af950f35381e9e530e/tensorflow/c/c_api.cc#L2272) rather than adding potential inputs to opts.input_map a couple lines above. That's what I was referring to in [this comment](https://github.com/tensorflow/tensorflow/issues/26371#issuecomment-477820714) in our previous thread ("I found a relatively simple way to add control inputs from the outer graph to the loop body (by only changing while loop code, not allowing control inputs between separate graphs in general)").

Thanks again for all your help Skye, I always appreciate it!

**Comment by skye** [comment]
Ah you're right, my suggestion didn't make much sense. What I'm trying to achieve is that the [ImportGraphDef call](https://github.com/tensorflow/tensorflow/blob/e9bd33a41d181593748d12af950f35381e9e530e/tensorflow/c/c_api.cc#L2282) allows `gdef` to reference nodes from the outer graph, instead of raising the "Unknown input node" error. I _think_ you can do this by populating `opts.input_map` with every outer node's name+index as both the key and the value. The keys of input_map are supposed to appear in gdef, not in the outer graph, but I don't think we actually check this (it would happen [here](https://github.com/tensorflow/tensorflow/blob/9d558ddba177859ebd7e4018a405f68dcc5e1836/tensorflow/core/graph/graph_constructor.cc#L435), although I could be missing something).

Does this make sense? It's kind of a hack, but I just wanna see if this works and then we can figure out how to make it less hacky, or if there are more fundamental problems we need to solve.

**Comment by melissagrueter** [comment]
Got it, thanks for clarifying! I will give that a try and report back.

**Comment by melissagrueter** [comment]
Hi @skye! 

I gave it a try and my AccessOuterOp unit test (see above) is now passing, but a few of the other tests are now failing:

[ FAILED ] CApiWhileLoopTest.NestedLoop
[ FAILED ] CApiWhileLoopTest.InvalidCondOutputNode
[ FAILED ] CApiWhileLoopTest.InvalidBodyOutputNode
[ FAILED ] CApiWhileLoopTest.WrongGraph

It seems to me that the InvalidCondOutputNode, InvalidBodyOutputNode, and WrongGraph tests would all be expected to fail now, since they're all looking for an error if you set the cond or body outputs to a node in the outer graph (and this hack is intended to prevent such an error). 

I'm not sure yet why this causes the NestedLoop test to fail. Here's the error NestedLoop produces:
{{node test_loop/body/inner_loop/Merge}} has inputs from different frames. The input {{node test_loop/body/inner_loop/NextIteration}} is in frame 'test_loop'. The input {{node test_loop/body/inner_loop/Enter}} is in frame 'inner_loop'.

**Comment by skye** [comment]
Oof, I bet NestedLoop is failing because there's a problem with my input_map hack: if both the outer and inner graph have a node with the same name, there's no way to distinguish if a given input refers to the outer or inner tensor. I'm guessing NestedLoop is running into this because the inner and outer loops will end up using the same node names. This wasn't a problem before because keys in the input_map always referred to the inner graph being imported.

This is unfortunate, because it means we'll probably need something more heavy-duty, like how Python FuncGraphs "capture" external tensors explicitly instead of just referring to them. We can maybe work around it for now by creating a unique name scope that we use when creating the while loop graphs... the C API doesn't have name scopes though :( This is even hackier, but can you create a name scope in Java that you use when creating the while loop graphs? Otherwise maybe we should think about how to properly refer to or "capture" external tensors in the C API.

**Comment by melissagrueter** [comment]
What if we just add a list of outer graph nodes to allow as a parameter to copygraph? There should be no ambiguity if they always refer to nodes in the outer graph, and we could presumably allow these as inputs the same way we allow the nodes that are values in the input_map.

**Comment by skye** [comment]
That might work for the particular use cases you have in mind, but there's still the fundamental problem that if you have an input "foo:0", an internal node "foo", and an external node "foo" (in the parameter whitelist), there's no way to determine which node "foo:0" is referring to. Having the whitelist makes you less likely to find yourself in this situation, but it's still possible.

**Comment by melissagrueter** [comment]
Hmm, let me back up a bit -- could you clarify why you believe this a naming issue? I don't quite see it yet.

The nodes that cause NestedLoop to fail seem to already be named uniquely depending on which graph they're in — their names all include inner_loop:

{{node test_loop/body/**inner_loop**/Merge}} has inputs from different frames. The input {{node test_loop/body/**inner_loop**/NextIteration}} is in frame 'test_loop'. The input {{node test_loop/body/**inner_loop**/Enter}} is in frame 'inner_loop'.

They also all appear to be in the same graph (in the inner_loop). If the problem was mixing up nodes in the inner and outer graph due to their names being the same, wouldn't we expect Merge and its inputs not to all be in the same graph?

**Comment by melissagrueter** [comment]
Update I am getting NestedLoop to pass now, just need to dig a bit more to make sure the fix can work in general... :)

**Comment by melissagrueter** [comment]
Hi @skye ,

As you could have predicted, my latest quick fix didn't work out — I understand the naming issue now. :) 

It seems to me that the best way forward is to go for the long-term fix and move toward functional control flow. As we've discussed [in another thread](https://github.com/tensorflow/tensorflow/issues/28385), I'm also interested in getting a cond op working. To that end, I'd like to avoid putting too much time into short-term while-loop-specific hacks, since a non-functional cond would presumably have the same issue as the current non-functional while. From what I'm seeing here and our discussions so far, my understanding is that the best way to make progress on both cond and while is to get something like FuncGraphs working in the C API. Do you agree?

If so, do you have an idea of what the steps are to getting functional control flow working in the C API?  We've talked about FuncGraphs as a good first step, but I'm not yet sure how big a chunk of the work toward achieving functional control flow FuncGraphs are, or if FuncGraphs depend on other things the C API doesn't support that I'd need to add first. 

If FuncGraphs are the best first step, I've been thinking it might make sense to proceed like this:

- Implement FuncGraphs for the C API
- Use FuncGraphs as the cond and body subgraphs of the existing non-functional while loop, just to get while up and running quickly and test the FuncGraphs (Is this possible? This is similar to what I think you were suggesting as a first step for cond in the [other thred](https://github.com/tensorflow/tensorflow/issues/28385))
- Implement functional cond (From looking at the functional control flow RFCs, it looks like functional cond is significantly easier to tackle than functional while, so it looks like a better place to start) 
- Implement functional while 

Let me know what you think. Any recommendations you have about how to go about tackling this would of course be incredibly helpful!

**Comment by melissagrueter** [comment]
One more question — I see that for the Python's functional control flow, there's a lowering pass to the non-functional cond and while. Is it critical to be able to lower functional control flow ops into their non-functional equivalents? In other words, will we still end up needing to fix the non-functional control flow even if we do implement functional control flow?

**Comment by skye** [comment]
Hi Melissa, once again sorry for leaving you hanging.

To answer your latest question, the lowering happens in the TF runtime, so you'll get it "for free" once functional cond + while are implemented. To actually answer your question, it's technically not necessary, but the non-lowered path isn't really tested so in practice you do wanna lower.

I'm wary to send you down the FuncGraph + functional control flow path because I'm worried it's a lot of work, and also will require some attention from TF API owners, who are perpetually very busy. But, it does seem like hacking this together isn't gonna work, so if you understand there's a risk this will stall, let's do it.

One thing to note up front -- it will be a lot more work get this all working with gradients. Are you gonna need to compute gradients through control flow?

Your overall plan sounds good. The key here will implementing FuncGraphs in the most minimal way possible to get functional control flow working (note that https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/func_graph.py is almost 1000 lines of Python... we don't want that). I think this means tracking inputs/outputs, and captures (which are represented as an input + an external Tensor that we always pass to that input). I think we should initially develop this as an experimental API in case we have to change it.

Before we dig into this too much, are you sure you want to continue? :) (and will you need gradients?) I'll see if I can get some internal feedback too.

**Comment by melissagrueter** [comment]
Hi Skye,

My goal is to get while and cond ops working, with gradients, in whatever is the most efficient (and also stable) way. 

Would the following more reasonable than my earlier plan to fully implement functional control flow? 

* Implement some minimal FuncGraph for the C API
* Use FuncGraphs as the cond/body subgraphs of the existing non-functional while loop
* Write a non-functional cond that just uses FuncGraphs as the branches

A couple of questions to help me understand whether the above would actually be less work than my previous plan: 

RE: gradients,

* If we use FuncGraphs as the cond/body subgraphs of the existing non-functional while, will the existing while loop gradient implementation just work? If not, what would it take to get the gradients working (and would it be the same amount of work as it would be to get gradients working for a functional while loop)?
    * In the case that the existing while loop gradient implementation just works, we still don't have the corresponding CondContext implementation for cond. Would it be simpler to implement CondContext in that case, or would it make more sense to just get the gradients working for the functional cond? 

Also, is there existing C API support for functions that will be relevant/helpful when getting FuncGraphs working? I've looked briefly at [c_api_function.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api_function.cc) — will any of that support for functions be useful?

Thanks again for all the help, Skye! :)

**Comment by melissagrueter** [comment]
Hi @skye! Just following up :)

**Comment by ewilderj** [comment]
cc @wicke this issue has gotten stuck and is blocking SIG JVM.

**Comment by martinwicke** [comment]
@saxenasaurabh could you take a look? I think sending people down the functional while path is now probably the right thing to do.

**Comment by saxenasaurabh** [comment]
Sorry this got pushed to the back-burner since we have been focusing on fixing bugs for 2.0. 
I am planning to find some time over the next weeks to design how FuncGraphs in the C API might look like. I think this will be important to pursue soon since we are starting to see issues in other areas e.g. C++ shape inference, colocations etc. because of missing nested graph information. Maybe I can send out a RFC so that the SIG JVM community can join in.

**Comment by melissagrueter** [comment]
Thanks @saxenasaurabh for following up! I'd love to hear what you come up with in terms of design -- an RFC sounds like a great idea!

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
